### PR TITLE
add support for macos app bundles / dirs retrieval at runtime

### PR DESCRIPTION
**Pull request by flxzt**
This add some configuration to make app bundles in macos work. addresses #41

### CODE REVIEW

**Comment by dehesselle** [comment]
Hi @flxzt ,
I've picked up work on the app again. For some reason I've had mixed results with this branch which I cannot really explain/reproduce, so before getting too deep into it, can I ask you to rebase this on current master?

One thing is already clear though: I forgot that there is one more change necessary. This needs to be made dynamic as well:
https://github.com/flxzt/rnote/blob/777fdc0132b2e5bed0df4ff1f70e783936e782a1/rnote-ui/src/app/mod.rs#L147
The resources file (`resources.gresource`) needs to be loaded from a location relative to the application binary as well. Since we're already setting `XDG_DATA_DIRS`, you can use that, e.g. `XDG_DATA_DIRS/rnote/resources.gresource`.
I am aware that this suggestion is risky as `XDG_DATA_DIRS` is supposed to contain a list of directories, but since that is something we control at this point (i.e. a list with one element), it would be safe. (Alternatively you'd have to iterate through the list of directories and try everyone of them to locate the `resources.gresource` file.)

File: rnote-ui/src/main.rs [hunk 1]
```diff
@@ -77,3 +82,32 @@ fn setup_windows_env() -> anyhow::Result<()> {
...
+            } else {
+                false
+            }
+        })
+    {
+        std::env::set_var("XDG_DATA_DIRS", &current_dir.join("/../Resources/share"));
+        std::env::set_var(
+            "GDK_PIXBUF_MODULE_FILE",
+            current_dir.join(PathBuf::from(
+                "/../Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/loaders.cache",
```
Comment (dehesselle) [inline rnote-ui/src/main.rs:108]: This is the wrong path. Lose the `loaders` subdirectory:
```
/../Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache
```

File: rnote-ui/src/main.rs [hunk 2]
```diff
@@ -77,3 +82,32 @@ fn setup_windows_env() -> anyhow::Result<()> {
...
+        .zip(current_dir.components().skip(1))
+        .any(|(a, b)| {
+            if let (Component::Normal(a), Component::Normal(b)) = (a, b) {
+                a == OsStr::new("Contents") && b == OsStr::new("MacOS")
+            } else {
+                false
+            }
+        })
+    {
+        std::env::set_var("XDG_DATA_DIRS", &current_dir.join("/../Resources/share"));
```
Comment (dehesselle) [inline rnote-ui/src/main.rs:104]: Apologies if my assumption is incorrect: I think `current_dir` does not contain the directory name but still the full path to the binary at this point. I'd expect a `skip()` here same as above.

File: rnote-ui/src/main.rs (hunk 1 above)
Comment (flxzt) [reply rnote-ui/src/main.rs:108]: ah right, probably copy-pasted the path from the win setup here

File: rnote-ui/src/main.rs (hunk 2 above)
Comment (flxzt) [reply rnote-ui/src/main.rs]: no, you were right.

**Comment by flxzt** [comment]
okay, *I think* I addressed the issues.

What will the result be of your efforts, a script that can be executed on macos to create the bundle?

**Comment by dehesselle** [comment]
Great, I'll test it again later today.

The result will be this: https://gitlab.com/dehesselle/rnote_macos
(Still work in progress, see https://gitlab.com/dehesselle/zim_macos for a polished example.)

I'm using GitLab as I have CI infrastructure there (and GitHub's CI  is sadly rather unstable for me).

**Comment by dehesselle** [comment]
Almost there. This is the output I'm getting (I have added two `println!()` statements for `data_dir` and `lib_dir`):
```
data_dir=/../Resources/share
lib_dir=/../Resources/lib

(Rnote:51502): GLib-GIO-CRITICAL **: 23:23:31.465: g_settings_schema_source_lookup: assertion 'source != NULL' failed

(Rnote:51502): GLib-GIO-CRITICAL **: 23:23:31.465: g_settings_schema_source_lookup: assertion 'source != NULL' failed
thread 'main' panicked at 'Could not load gresource file: Error { domain: g-file-error-quark, code: 4, message: "Failed to open file “/../Resources/share/rnote/resources.gresource”: open() failed: No such file or directory" }', rnote-ui/src/app/mod.rs:155:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by flxzt** [comment]
argh, I don't know why that slash landed there. The executable name doesn't appear in your print output, but it is apparently lost when joining with the (invalid) absolute path (see [here](https://github.com/rust-lang/rust/issues/16507) )

**Comment by dehesselle** [comment]
Sorry you have to check this once more, now it wants to open
```
MacOS/../Resources/share/rnote/resources.gresource
```
This needs to be the full canonical path.

**Comment by flxzt** [comment]
The mention of "dir name" [here](https://github.com/flxzt/rnote/issues/41#issuecomment-1401114485) and the comment [here](https://github.com/flxzt/rnote/pull/470#discussion_r1130252110) was confusing.

**Comment by dehesselle** [comment]
Apologies for the confusion, the intention was to "think like coreutils", e.g. `dirname /whatever/path/to/some/random/binary`

**Comment by flxzt** [comment]
no worries, I also read it wrong in the sense of "folder name" .

**Comment by dehesselle** [comment]
Awesome, Rnote works now out of the box as macOS app! 🎉 

I'll have to do some more polishing on my end, but you can basically merge this to master and with 0.5.17 (or whatever comes next) you have a macOS app.

One more thing though, which is either for me to figure out or it might be related to what we've been doing here: can you tell me from where the logo in the about screen is  supposed to come from?
image

**Comment by flxzt** [comment]
Thanks a lot for doing this!

I want to mention and link to it in the README & on the website, do you have any suggestions for that / a formulation for macos users how to install / use it?

> can you tell me from where the logo in the about screen is supposed to come from?

That should be the icon which gets installed to `<data dir>/icons/hicolor/scalable/apps/com.github.flxzt.rnote.svg`

**Comment by dehesselle** [comment]
You can link to the latest release by using
```
https://gitlab.com/dehesselle/rnote_macos/-/releases/permalink/latest
```
This doesn't work yet since I haven't created a release yet. I can create a release this weekend if you are good to go on 0.5.17.

I'll put everything that people need to know in the readme, so for that you can link to the GitLab repo (see [zim_macos](https://gitlab.com/dehesselle/zim_macos) as example).

You can add it to your readme/website as you see fit. My first guess is to add it to "Installation" below Flathub? "Rnote is available for macOS here:".

**Comment by dehesselle** [comment]
@flxzt final addendum: first release is live!

**Pull request merged**